osu.Game/Overlays/Profile/UserGraph.cs

```diff
@@ -0,0 +1,276 @@
+﻿// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using osu.Framework.Allocation;
+using osu.Framework.Graphics;
+using osu.Framework.Graphics.Containers;
+using osu.Framework.Graphics.Cursor;
+using osu.Framework.Graphics.Shapes;
+using osu.Framework.Input.Events;
+using osu.Game.Graphics;
+using osu.Game.Graphics.Sprites;
+using osu.Game.Graphics.UserInterface;
+using osuTK;
+
+namespace osu.Game.Overlays.Profile
+{
+    public abstract class UserGraph<TKey, TValue> : Container, IHasCustomTooltip
+    {
+        protected const float FADE_DURATION = 150;
+
+        private readonly UserLineGraph graph;
+        private KeyValuePair<TKey, TValue>[] data;
+        private int dataIndex;
+
+        protected UserGraph()
+        {
+            data = Array.Empty<KeyValuePair<TKey, TValue>>();
```

Comment:
i don't understand why internally this thing makes this non-nullable while handling incoming nulls. usually we support null or don't the whole way, not at point of `set`. i'd either make `data` nullable or stop supporting null as a valid input.

Reply:
that one was me - think there was a place where null was passed in (might have been a test) so I wanted to do some defensive programming but it does make sense to stick to one convention

Reply:
fine to do this, but i'd move to the field definition and clearly mark as `[NotNull]`